### DISCUSSION TITLE
Custom variant has invalid format string

### DISCUSSION THREAD

**Discussion by cossssmin** [tailwindlabs/tailwindcss#8647] [Help]
I register [this variant](https://github.com/maizzle/tailwindcss-email-variants/blob/a2998eabc2cf47dde3768993f2138a36790b8ca2/src/index.js#L28-L32) in the [tailwindcss-email-variants](https://github.com/maizzle/tailwindcss-email-variants) plugin:

```js
const get = require('lodash.get')
const plugin = require('tailwindcss/plugin')

module.exports = plugin.withOptions(
  function (options) {
    addVariant('outlook-web', ctx => {
      const foo = get(ctx.container.nodes[0], 'raws.tailwind.classCandidate', '&')
      return `[class~="x_outlook-web\\:${foo}"]`
    })
  }
)
```

So that for example `outlook-web:hidden` would result in this selector: `[class~="x_outlook-web\:hidden"]`

Worked fine with `3.0.x`, but now with latest `3.1.x` I'm getting this error:

```
Your custom variant `outlook-web` has an invalid format string. Make sure it's an at-rule or contains a `&` placeholder.
```

You may see the (now failing) test here:

Test: https://github.com/maizzle/tailwindcss-email-variants/blob/a2998eabc2cf47dde3768993f2138a36790b8ca2/src/index.test.js#L114-L126

CI run: https://github.com/maizzle/tailwindcss-email-variants/runs/6900405408?check_suite_focus=true

The plugin also offers the ability to register your own variants, that you may pass as an options object, so [this](https://github.com/maizzle/tailwindcss-email-variants/blob/a2998eabc2cf47dde3768993f2138a36790b8ca2/src/index.js#L37-L40) is also failing now:

https://github.com/maizzle/tailwindcss-email-variants/blob/a2998eabc2cf47dde3768993f2138a36790b8ca2/src/index.test.js#L156-L165

Is there any way to get those working again?

**Comment by huyenltnguyen** [comment]
Would it work if you add a `&` prefix to the class?

```diff
module.exports = plugin.withOptions(
  function (options) {
    addVariant('outlook-web', ctx => {
      const foo = get(ctx.container.nodes[0], 'raws.tailwind.classCandidate', '&')
-      return `[class~="x_outlook-web\\:${foo}"]`
+      return `[class~="[&x_outlook-web\\]:${foo}"]` 
    })
  }
)
```

**Reply by cossssmin** [reply]
No because that'd output `[class~=".outlook-web\:hiddenx_outlook-web\:hidden"]`.

Closest I could get was `[class~="x_&"]`, but that still outputs the dot i.e. `[class~="x_.outlook-web\:hidden"]`

**Reply by cossssmin** [reply]
And that still doesn't solve the user-defined variants that I was registering with:

```js
// User-defined variants
Object.keys(userVariants).forEach(key => {
  addVariant(key, userVariants[key])
})
```
